plugins/outputs/cloudwatch/cloudwatch.go

```diff
@@ -279,48 +276,90 @@ func (c *CloudWatch) timeToPublish(b *MetricDatumBatch) bool {
 	return len(b.Partition) > 0 && time.Now().Sub(b.BeginTime) >= c.ForceFlushInterval.Duration
 }
 
+// getFirstPushMs returns the time at which the first upload should occur.
+// It uses random jitter as an offset from the start of the given interval.
+func getFirstPushMs(interval time.Duration) int64 {
+	publishJitter := publishJitter(interval)
+	log.Printf("I! cloudwatch: publish with ForceFlushInterval: %v, Publish Jitter: %v",
+		interval, publishJitter)
+	nowMs := time.Now().UnixMilli()
+	// Truncate i.e. round down, then add jitter.
+	// If the rounded down time is in the past, move it forward.
+	nextMs := nowMs - (nowMs % interval.Milliseconds()) + publishJitter.Milliseconds()
```

Comment:
Maybe I'm wrong but the first publish will off the the time for `ForceFlushInterval` and will cause wasteful API call if the random falls into first half of ` ForceFlushInterval`. Therefore, would be good to always add `interval.Milliseconds()/2` to allow the Agent to collect more metrics? If it's full then good its not wasteful

Reply:
My thinking was that I don't want to make the Agent ever wait longer than the given `ForceFlushInterval`.
So if current time is `04:26:04`, `ForceFlushInterval` is 60s, and jitter is `55s`, I do not want the agent waiting until `04:27:55. I'd prefer to use `04:26:55`.

There is no wasteful API call since if there are no metrics in any batches, the call does nothing.

Reply:
> I do not want the agent waiting until 04:27:55. I'd prefer to use 04:26:55

This only happens with initial push so I understand your concern. However, would the  jitter is `2s` be the number of collected metrics as 55s? However, after your confirmation, then its good to me

```
There is no wasteful API call since if there are no metrics in any batches, the call does nothing.
```

Reply:
Correct, calling `c.pushMetricDatumBatch()` does this:
```
func (c *CloudWatch) pushMetricDatumBatch() {
	for {
		select {
		case datumBatch := <-c.datumBatchChan:
			c.publisher.Publish(datumBatch)
			continue
		default:
		}
		break
	}
}
```
So if there are no batches in the channel it does nothing.

And if you take a look at: `func (c *CloudWatch) pushMetricDatum() {` you will see it puts a batch on the channel if the batch is full or if the batch has existed for `>= c.ForceFlushInterval.Duration`.